Mocking callbacks blog (#1288)
stoplightio/stoplight.io: netlify/blog-posts/mocking-callbacks-openapi-prism.md

---
path: /blog/mocking-callbacks-openapi-prism
tags:
  - blog
  - blog-design
  - blog-mocking
  - blog-featured
  - blog-general
  - blog-industry
relatedTags:
  - blog 
  - blog-design
  - blog-mocking
  - blog-featured
  - blog-general
  - blog-industry
publishedDate: 2019-12-12
author: Karol Maciaszek
title: Mocking Callbacks with OpenAPI and Prism

> subtitle: What are callback, why use them, and how to set them up using OpenAPI and Prism. 

Reviewer: ```suggestion
subtitle: What are callbacks, why use them, and how to set them up using OpenAPI and Prism. 
```

image: /images/rotary-phone.jpg
color: black
includeToc: true
actionBar:
  ctas:
  enabled: false
meta:
  canonical: https://11sigma.com/blog/2019-11-22--mocking-callbacks/#how-to-set-up-callbacks-with-openapi-and-prism
  description: What are callback, why use them, and how to set them up using OpenAPI and Prism.
  favicon: /images/mark_light_bg.png
  robots: 'index, follow'
  title: Mocking Callbacks with OpenAPI and Prism | Stoplight Blog
  image: /images/rotary-phone.jpg
  twitter:

>     description: What are callback, why use them, and how to set them up using OpenAPI and Prism.

Reviewer: ```suggestion
    description: What are callbacks, why use them, and how to set them up using OpenAPI and Prism.
```

    title: Mocking Callbacks with OpenAPI and Prism | Stoplight API Corner
    image: /images/rotary-phone.jpg
    username: '@stoplightio'
---

> Callbacks are outgoing, asynchronous requests that service will make to some other service in response to an action. They are very similar to callback functions in popular programming languages like JavaScript. However, they are executed at the API abstraction level.

Reviewer: ```suggestion
Callbacks are outgoing, asynchronous requests that one service will make to some other service in response to an action. They are very similar to callback functions in popular programming languages like JavaScript. However, they are executed at the API abstraction level.
```

Author: I was originally going to leave grammar as-is, since this is canonical, but it probably doesn't hurt to spiff up a bit.

Reviewer: Ah, that makes sense! Sorry did not read closely. Let's let it be! I'll go back and undo my comments.

Callbacks can be observed in everyday life for instance when you order a package. If you subscribe to updates, the logistics company that delivers your parcel will send you notifications about events such as: “your package is ready to be picked up“, “your package arrived at its final destination“.

This type of communication happens asynchronously and in real-time. Very cool stuff!

## Other Approaches

Callbacks solve the asynchronous communication problem. Usually, systems performing computational-heavy operations use message-queues like Kafka, RabbitMQ or JMS-based. Selective subscription to events is implemented with topics. However, since the underlying protocol is not web-friendly (not HTTP-based) these solutions can't be leveraged by typical web applications.

> Another common approach is to use configure WebHooks. GitHub or Stripe expose a UI for setting them up.

Reviewer: Noob question: is 'configure WebHooks' its own noun? or should it read either: 

Another common approach is to configure WebHooks

Another common approach is to use WebHooks

Author: I also don't know - going to ask in slack real quick.

Reviewer 2: it can be either of your suggestions but not use configure

Did I say WebHook? Quick explanation: WebHook is the caller's API endpoint that receives callback requests.

The disadvantage of such a solution is that you need to manually update an external service (like GitHub) with the URL of the WebHook. That makes your configuration decentralized, difficult to maintain and update.

## Why callbacks

- they are standardized in OpenAPI (see [documentation](https://swagger.io/docs/specification/callbacks/) and [specification](https://spec.openapis.org/oas/v3.0.2#callbackObject))
- it's a Web Standard meaning, it's designed to use it safely in an HTTP-based environment, often publicly available (client-facing)
- supports selective data subscription

> - reception endpoints configuration is kept on the caller side which makes it single source of truth

Reviewer: ```suggestion
- reception endpoints configuration is kept on the caller side which makes it the single source of truth
```

## How to set up callbacks with OpenAPI and Prism

Now let's set a goal. The goal is to mock a complete, real-life example of a callback hitting a WebHook. We will use Prism as a mocking tool.

> Let's assume there is a logistics company named Ship-n-Pray Ltd. As a partner of the company, we decided to integrate with their API to automate our processes. For example, when a package arrives, we would like to send an email to our clients. Or keeping track and statistics of how the logistics company operates.

Reviewer: ```suggestion
Let's assume there is a logistics company named Ship-n-Pray Ltd. As a partner of the company, we decided to integrate with their API to automate our processes. For example, when a package arrives, we would like to send an email to our clients. Or, keep track of statistics around how the logistics company operates.
```

        content:
          application/json:
            schema:
              $ref: '#/components/schemas/notification'
      responses:
        '200':
          description: 'Notification successfully processed'
          content:
            text/plain:
              examples:
                ok:
                  value: 'ok'
components:
  securitySchemes:
    TokenSecurty:
      type: apiKey
      name: token
      in: query
  schemas:
    parcelId:
      type: string
      pattern: '^[0-9]{16}$'
    notification:
      type: object
      properties:
        parcelId:
          $ref: "#/components/schemas/parcelId"
        status:
          type: string
          enum:
          - pick-up
          - in-transit
          - delivered
      required:
      - parcelId
      - status
```

Our API specification defines a single endpoint that will handle notifications reception. Simple as that!

## Let's Mock

We've defined specifications of logistics service API and our API. It's time to spin up `prism` instances and see some action.

The following command will run `prism` instance for the logistics company on `4010` port.

``` yaml
prism mock -p 4010 logistics.oas3.yaml
```

> Now please open new console and run following command:

Reviewer: ```suggestion
Now please open a new console and run the following command:
```
